File: app/src/main/cpp/ufs.cpp
#include "ufs.h"

UFS::USF(int N_): N(N_) {
	std::vector<int>(N,-1).swap(parents);
}

int UFS::find(int s) const {
	if(parents[s]<0) return s;
	else return find(parents[s]);
}

void UFS::get(std::vector<int>& rs) const {
	std::vector<int>().swap(rs);
	for(int i=0; i<N; ++i) if(parents[i]<0) rs.push_back(i);
}

void UFS::join(int s1, int s2) {
	int r1 = find(s1);
	int r2 = find(s2);

	if(r1==r2) return;
	if(parents[r2]<parents[r1]) parents[r1] = r2;
	else {
		if(parents[r1]==parents[r2]) --parents[r1];
		parents[r2] = r1;
	}
}

File: app/src/main/java/cn/edu/sjtu/iiot/system/batchqr/model/UFS.java
package cn.edu.sjtu.iiot.system.batchqr.model;

import java.util.ArrayList;

/**
 * Created by SHIYONG on 2018/3/26.
 */

public class UFS {
    public int[] set;

    public UFS(int nb_elements){
        this.set = new int[nb_elements];
        for(int i=0; i<nb_elements; i++){
            set[i] = -1;
        }
    }

    public int find(int x){
        if(set[x] < 0){
            return x;
        }
        else {
            return this.find(set[x]);
        }
    }

    public void join(int x1,int x2){
        int r1 = find(x1);
        int r2 = find(x2);

        if(set[r2]<set[r1]){
            set[r1] = r2;
        }else {
            if(set[r2] == set[r1]){
                set[r1]--;
            }
            set[r2] = r1;
        }
    }
}
